### PR TITLE
Kernel/PCI: Don't hold spinlocks when doing fast device enumeration

### PR DESCRIPTION
**Pull request by supercomputer7**
Instead, hold the lock while we copy the contents to a stack-based
Vector then iterate on it without any locking.

cc @awesomekling

### CODE REVIEW

File: Kernel/Bus/PCI/Access.cpp [hunk 1]
```diff
@@ -110,18 +110,28 @@ UNMAP_AFTER_INIT bool Access::initialize_for_one_pci_domain()
...
+        SpinlockLocker scan_locker(m_scan_lock);
+        auto domain_number = controller->domain_number();
+
+        VERIFY(!m_host_controllers.contains(domain_number));
+        // Note: We need to register the new controller as soon as possible, and
+        // definitely before enumerating devices behing that.
+        m_host_controllers.set(domain_number, move(controller));
+        m_host_controllers.get(domain_number).value()->enumerate_attached_devices([&](DeviceIdentifier const& device_identifier) -> void {
+            m_device_identifiers.append(device_identifier);
+            device_identifiers_behind_host_controller.append(device_identifier);
```
Comment (awesomekling) [inline Kernel/Bus/PCI/Access.cpp:128]: Note that `Vector::append()` will panic if we run out of memory.

File: Kernel/Bus/PCI/Access.cpp [hunk 2]
```diff
@@ -149,9 +159,15 @@ UNMAP_AFTER_INIT void Access::rescan_hardware()
...
-    SpinlockLocker locker(m_access_lock);
-    VERIFY(!m_device_identifiers.is_empty());
-    for (auto const& device_identifier : m_device_identifiers) {
+    // Note: We hold the m_access_lock for a brief moment just to ensure we get
+    // a complete Vector in case someone wants to mutate it.
+    Vector<DeviceIdentifier> device_identifiers;
+    {
+        SpinlockLocker locker(m_access_lock);
+        VERIFY(!m_device_identifiers.is_empty());
+        device_identifiers = m_device_identifiers;
```
Comment (awesomekling) [inline Kernel/Bus/PCI/Access.cpp:168]: Copying a vector like this will panic if we are out of memory.

**Comment by supercomputer7** [comment]
Rebased and force-pushed to re-test CI now it's working again and also to fix a merge conflict after #12445 was merged :)

**Comment by Panky-codes** [comment]
I noticed that booting from NVMe drive is currently broken in the master. After doing `git bisect`, I noticed that this was the offending commit:
64254b5df8c2 ("Kernel: Protect PCI access with spinlock instead of mutex") by @awesomekling 

There seems to be a deadlock situation in `StorageManagement.cpp` when we enumerate the devices because of the spinlock.   

I **tested** this patch and deadlock doesn't occur anymore (as we limit the scope of the spinlock) and I am able to boot with a NVMe drive again. 
To reproduce the issue, run:
```
SERENITY_NVME_ENABLE=1 Meta/serenity.sh run i686 "root=/dev/nvme0n1"
```

**Comment by creator1creeper1** [comment]
You could copy the contents into a VLA in order to not have any fallible allocation, which is a pretty weird thing to have when you just want to iterate over something.

**Comment by supercomputer7** [comment]
> You could copy the contents into a VLA in order to not have any fallible allocation, which is a pretty weird thing to have when you just want to iterate over something.

We are not using VLAs in the codebase. This is a trade-off between holding the lock when enumerating the PCI device list, or creating a copy which can be freely used without a lock.

**Comment by creator1creeper1** [comment]
Yes, I was just proposing another way to make a copy which would not require heap allocation. But if VLAs are generally frowned upon here, I guess it's not an option.

**Comment by supercomputer7** [comment]
Rebased after resolving a merge conflict.

**Comment by supercomputer7** [comment]
Rebased after resolving a merge conflict.

**Pull request merged**